planner, executor: fix the mysql compatibility for recursive CTE throws warning instead of Error for data truncation under strict mode

Reminiscent opened the pull request:
Enforce strict SQL mode for data truncation in CTEs, and add a test case.

<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: close https://github.com/pingcap/tidb/issues/64711

Problem Summary:
Recursive CTE truncation handling in TiDB was inconsistent with MySQL in a few important paths.

In MySQL, once a recursive CTE is materializing rows for the recursive worktable under strict SQL mode, truncation on that recursive producer path is treated as an error. In TiDB, several parts of the pipeline could still evaluate with ordinary SELECT warning semantics, so truncation that should surface as ERROR 1406 could instead be folded away
during planning, downgraded to a warning, or evaluated under an unscoped session context.

The gaps were spread across multiple layers:

- Planner-side expression rewriting could constant-fold recursive output expressions too early and hide execution-time truncation errors.
- Recursive UNION output alignment could miss the final cast/error boundary, especially when the recursive side was already a LogicalProjection.
- Scoped truncate handling did not propagate consistently into helper contexts derived from the session, such as PlanCtx / DistSQL / Ranger / BuildPB / pushdown flags.
- DistSQL scoped contexts initially had correctness issues around late-bound fields, shared statement-level coordination state, and timezone propagation.
- Detach logic for executors using wrapped ExprCtx / EvalCtx could stop freezing contexts into static snapshots, which risks detached execution reading mutable session state from a later statement.
- Recursive nested set-op operands could preserve ORDER BY / LIMIT syntax without applying the same recursive-query restriction checks as ordinary SELECT recursive members.

### What changed and how does it work?

This change introduces a scoped ExprCtx override model for recursive CTE truncate handling and threads it through the planner/executor stack in the places that actually participate in recursive materialization semantics.

1. Expression-context override infrastructure

A truncate-level override is now modeled as a scoped ExprCtx / EvalCtx wrapper instead of mutating the session-owned StmtCtx globally. This keeps the change local to the recursive CTE build/evaluation path and avoids leaking strict truncate behavior to unrelated parts of the statement.

The wrapper also preserves static-convertible behavior for both ExprContext and EvalContext, so later detach/freeze paths can still convert them into static snapshots.

2. Session-scoped helper contexts derived from ExprCtx

A new generic WithExprCtx / WithTruncateErrLevel session wrapper is used to make all directly derived helper contexts observe the same scoped expression semantics:

- GetPlanCtx()
- GetBuildPBCtx()
- GetRangerCtx()
- GetDistSQLCtx()
- pushdown flags

This is important because recursive CTE truncate handling is not only local expression evaluation. Range building, PB building, DistSQL request building, and pushed-down execution must agree on the same truncate/error policy.

For DistSQL specifically, the scoped wrapper now:

- keeps one shared detached DistSQLContext per scoped session so statement-level coordination state such as TryCopLiteWorker remains shared across readers,
- refreshes late-bound fields such as RunawayChecker and RUV2Metrics on every access,
- mirrors the scoped ErrCtx and timezone into DistSQL so pushed-down evaluation stays consistent with the scoped ExprCtx.

GetPushDownFlags() was also updated to derive truncate/error-related bits from the active scoped ExprCtx while preserving the original statement-kind bits from StmtCtx, matching the old contract but with scoped truncate semantics.

3. Planner-side recursive output handling

Planner-side handling is intentionally narrower than executor-side MySQL behavior.

MySQL’s prepare/optimizer phase mainly folds conditions (WHERE / HAVING / join conditions), not recursive output materialization itself. TiDB, however, could fold recursive output expressions early enough to hide execution-time truncation errors. To avoid that mismatch, planner-side strict truncate handling is applied only where TiDB must
preserve the later materialization error boundary:

- top-level recursive output projection,
- recursive UNION schema-alignment projection.

That means:

- recursive output expressions are rewritten under scoped strict truncate handling,
- final recursive-side alignment to the seed-defined CTE schema also uses the scoped strict context,
- WHERE / ON / HAVING keep their normal planner folding behavior.

This keeps the planner fix narrow and aligned with the actual problem: preventing TiDB from swallowing recursive materialization errors before execution.

4. Recursive UNION alignment and LogicalProjection handling

buildProjection4CTEUnion was fixed to align the recursive side against the final CTE schema more carefully.

The key issue is that Schema() can look aligned while the real projection expressions still do not contain the final cast boundary. For a recursive side that is already a LogicalProjection, the code now checks and fixes the projection expressions themselves instead of relying only on schema columns. This ensures the final recursive worktable
alignment still builds the required cast and still reports truncation as an error in strict mode.

5. Recursive set-op operand validation

When recursive members contain nested set-op operands represented as SetOprSelectList, ORDER BY / LIMIT is now rejected with the same recursive-query restriction as ordinary recursive SELECT members. This prevents nested recursive set operands from bypassing the existing recursive CTE syntax guard after preserving their parser-attached OrderBy / Limit.

6. Executor-side MySQL-compatible recursive producer behavior

On the executor side, seed and recursive subplans are built under a scoped strict truncate context when the recursive CTE runs in strict SQL mode. This matches MySQL’s
recursive materialization behavior, where the recursive producer path runs with strict SELECT error handling.

The scope is intentionally limited to child subplan construction:

- the built child executors capture the scoped ExprCtx / DistSQL / pushdown state they need,
- cteProducer itself keeps the original session context for runtime bookkeeping such as trackers, recursion depth, and DISTINCT deduplication state that continues to read the statement-owned StmtCtx.

7. Detach correctness for wrapped contexts

Executor detach logic was generalized from concrete sessionexpr types to static-convertible capability checks, so wrapped contexts created by truncate scoping still detach correctly.

This fixes detach behavior for:

- table/index readers using wrapped ExprContext,
- index lookup nested reader context,
- projection/selection executors that store EvalContext directly.

As a result, detached cursors/background result sets continue to freeze recursive-CTE strict contexts into static snapshots instead of accidentally retaining live session-backed context.

8. Tests

The patch adds and updates targeted regression coverage for:

- recursive CTE strict truncate behavior,
- recursive UNION alignment,
- nested set-op recursive syntax validation,
- scoped session helper contexts,
- DistSQL scoped context sharing, late-bound refresh, and timezone propagation,
- detach behavior for wrapped ExprCtx / EvalCtx.

These tests were verified both before and after the fixes where applicable, to ensure the new coverage actually catches the bugs being fixed.

### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > <!-- Or your custom  "No need to test" reasons -->

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved string truncation handling in recursive CTEs and related projections/subqueries: strict SQL mode now yields Error 1406 where appropriate; non-strict mode emits warnings and returns truncated results. Prepared-statement execution paths also follow the same behavior.

* **Tests**
  * Added extensive integration tests covering recursive CTE truncation across SQL modes, prepared statements, EXPLAIN/warnings checks, and updated lateral-join recursive test expectations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Reminiscent committed:
fix: Enforce strict SQL mode for data truncation in CTEs, correct recursive plan stats, and add a test case.

fix and add more test cases

fix make_check

fix

fix

fix

fix

fix

fix

fix

Reminiscent commented:
/retest

Reminiscent commented:
/retest

lcwangchao commented on pkg/planner/core/operator/logicalop/logical_cte.go:
```diff
@@ -227,6 +236,18 @@ func (p *LogicalCTE) DeriveStats(_ []*property.StatsInfo, selfSchema *expression
 
+// MakeCTEStrictTruncateErrPlanCtx returns a wrapped PlanContext so CTE worktable writes treat truncation like INSERT
+// in strict SQL mode, while keeping the behavior scoped to the CTE subtree.
+func MakeCTEStrictTruncateErrPlanCtx(pctx base.PlanContext) base.PlanContext {
```
Could we make it as a common function `planctx.WithTruncateErrLevel(level)` and move it to planctx package?

lcwangchao commented on pkg/expression/column.go:
```diff
@@ -454,20 +454,35 @@ const columnPrefix = "Column#"
+		UnwrapEvalContext() EvalContext
+	}
+
+	for range 8 {
```
Also, could we introduce a method of `UnwrapInternalSctx` to reduce this kind of loop?

Reminiscent committed:
rename

lcwangchao commented on pkg/executor/builder.go:
```diff
@@ -5990,9 +5937,23 @@ func (b *executorBuilder) buildCTE(v *physicalop.PhysicalCTE) exec.Executor {
+			b.ctx = buildCtx
+			defer func() { b.ctx = origCtx }()
+			return b.build(plan)
+		}
```
Is it better? Seems easier to understand, especially for normal path

```suggestion
		buildSubPlan = b.build
		// MySQL evaluates a recursive CTE by writing rows into an internal worktable, so truncation is handled like INSERT.
		// In strict SQL mode, build the seed/recursive sub-plans with a strict ExprCtx only.
		if v.RecurPlan != nil && origCtx.GetSessionVars().SQLMode.HasStrictMode() {
			buildCtx := wrapSessionCtxForCTEStrictTruncateErr(origCtx)
			buildSubPlan := func(plan base.Plan) exec.Executor {
				origCtx := b.ctx
				b.ctx = buildCtx
				defer func() { b.ctx = origCtx }()
				return b.build(plan)
			}
		}


```

lcwangchao commented on pkg/executor/select.go:
```diff
@@ -610,6 +610,9 @@ func init() {
 		}
+		// Ensure expression evaluation during optimization (e.g. uncorrelated scalar subquery preprocessing)
+		// follows the same ExprCtx as the PlanContext, even if the PlanContext is a wrapper.
+		sctx = wrapSessionCtxWithExprCtx(sctx, pctx.GetExprCtx())
```
Does it affect the previous behavior?

Reminiscent committed:
address comments

Reminiscent commented on pkg/planner/core/plan.go:
```diff
@@ -30,11 +30,35 @@ import (
+		}
+		pctx = next
 	}
-	return sctx, nil
```
Update. PTAL

Reminiscent committed:
update

Reminiscent commented:
@lcwangchao Updated. Add the assert and return error when there are some new planContext wrappers forget to implement the `UnwrapAsInternalSctx` function in the future.

Reminiscent commented:
/retest

Reminiscent committed:
address yiding's comments

Reminiscent committed:
address comments

Reminiscent committed:
add more comments to explain the usage

Reminiscent commented on pkg/planner/planctx/context.go:
```diff
@@ -82,6 +83,11 @@ type Common interface {
+	// UnwrapAsInternalSctx returns the underlying internal session context as `any`.
+	//
+	// NOTE: It returns `any` instead of `sessionctx.Context` to avoid a dependency cycle
+	// (sessionctx -> planctx). Callers can type-assert the returned value when needed.
```
Added.

XuHuaiyu commented on pkg/planner/core/logical_plan_builder.go:
This changes the `ExprCtx` during recursive CTE plan building, not only during recursive worktable materialization. That means planning-time paths such as constant folding, `EXPLAIN`, `PREPARE`, or optimizer-time subquery evaluation can now observe `1406` as well. Could we confirm MySQL behavior for those non-execution entry points too, or otherwise narrow the override to the execution/materialization path only?

Reminiscent committed:
resolve conflicts and add more test cases

Reminiscent committed:
tests/integrationtest: add recursive CTE explain and prepare coverage

Reminiscent commented:
> One higher-level concern with this fix: the new override changes the `ExprCtx` during recursive CTE plan building, not only during recursive worktable materialization. Because of that, planning-time paths such as constant folding, `EXPLAIN`, `PREPARE`, or optimizer-time subquery evaluation may now observe `1406` as well.
> 
> Could we confirm MySQL behavior for those non-execution entry points too, or otherwise narrow the override to the execution/materialization path only? Right now the PR description reads as if the behavior change is scoped to recursive CTE execution, but the current wiring seems a bit broader.

I verified the non-execution entry points on MySQL 8.4.7 and added e2e coverage on the TiDB side.

For the MySQL case:

```sql
SET @@SESSION.sql_mode = 'STRICT_TRANS_TABLES';
EXPLAIN FORMAT=TREE WITH RECURSIVE cte AS (
  SELECT CAST('abc' AS CHAR(3)) AS s, 1 AS n
  UNION ALL
  SELECT 'abcdefgh', n + 1
  FROM cte
  WHERE n < 2
)
SELECT * FROM cte;
SELECT @@warning_count AS warning_count;
```

MySQL 8.4.7 does not raise an error there, and warning_count is 0.

Likewise, PREPARE succeeds without warning, and the error is raised only on EXECUTE:

```sql
SET @@SESSION.sql_mode = 'STRICT_TRANS_TABLES';
PREPARE stmt FROM "WITH RECURSIVE cte AS (
  SELECT CAST('abc' AS CHAR(3)) AS s, 1 AS n
  UNION ALL
  SELECT 'abcdefgh', n + 1
  FROM cte
  WHERE n < 2
) SELECT * FROM cte";
SELECT @@warning_count AS warning_count_after_prepare;
EXECUTE stmt;
```

The MySQL result is:

- warning_count_after_prepare = 0
- EXECUTE raises ERROR 1406 (22001): Data too long for column 's' at row 1

I added matching e2e coverage in tests/integrationtest/t/executor/cte.test on the TiDB side. The current TiDB branch behaves the same on these entry points:

- EXPLAIN succeeds, @@warning_count = 0
- PREPARE succeeds, @@warning_count = 0
- EXECUTE raises 1406

## On why the planner-side change is still needed:

The planner-side override is not there to make EXPLAIN / PREPARE fail. It is needed to prevent planning-time silent truncation from changing the runtime semantics of EXECUTE.

Concretely, recursive CTEs are syntactically under a SELECT, and TiDB's normal SELECT statement context uses truncate-as-warning semantics. Without a scoped override, build-
time expression evaluation in the recursive seed / recursive member can inherit plain-SELECT behavior.

That matters because some expressions may be evaluated or simplified during planning/optimization:

- constant folding may try to evaluate constant casts/coercions
- uncorrelated scalar subqueries may also have planning-time evaluation paths

If those planning-time paths run under the default SELECT truncate policy, a too-long value can be silently rewritten into an already-truncated constant before execution.
Once that happens, runtime never sees the original overlong value anymore, so EXECUTE incorrectly succeeds.

I verified this locally by temporarily removing only the planner-side WithTruncateErrLevel(..., LevelError) in buildCte(): after that change, EXPLAIN and PREPARE still did
not fail, but EXECUTE no longer raised an error either and instead returned silently truncated rows (abc, abc) in strict mode.

## On why EXPLAIN / PREPARE still do not fail even with the planner-side override:

This is not because errors are blindly swallowed. There are two separate reasons:

1. EXPLAIN / PREPARE do not materialize the recursive worktable, so they do not hit the runtime write path.
2. For planning-time simplification, constant folding does not turn evaluation errors into statement errors. If folding a constant expression returns an error, TiDB keeps the
   original expression instead of failing the statement immediately. For EXPLAIN, non-correlated scalar subqueries may also be kept as unevaluated ScalarSubQueryExpr
   placeholders instead of being executed.

So the planner-side strict ExprCtx is needed to stop silent build-time truncation, but it does not imply that EXPLAIN / PREPARE themselves should fail.

Reminiscent commented on tests/integrationtest/t/executor/cte.test:
```diff
@@ -153,3 +153,59 @@ create table t1(c1 int);
 insert into t1 values(0), (1), (2), (3), (4), (5), (6), (7), (8), (9), (10), (11), (12), (13), (14), (15), (16), (17), (18), (19), (20), (21), (22), (23), (24), (25), (26), (27), (28), (29), (30), (31), (32), (33), (34), (35), (36), (37), (38), (39), (40), (41), (42), (43), (44), (45), (46), (47), (48), (49), (50), (51), (52), (53), (54), (55), (56), (57), (58), (59), (60), (61), (62), (63), (64) ...
 with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from cte1 limit 1 offset 100) select * from cte1;
 set tidb_max_chunk_size=default;
+
```
Added the explain and prepare test cases

Reminiscent commented:
/hold

Reminiscent committed:
*: align recursive CTE truncate handling with scoped ExprCtx
